File: app/src/main/java/com/timsimonhughes/atlas/network/RetrofitClientInstance.java
package com.timsimonhughes.atlas.network;


import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;

import java.io.IOException;

import okhttp3.Cache;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class RetrofitClientInstance {

    public static Retrofit getRetrofitInstance(Context context, int cacheSize) {

        Retrofit retrofit;

        Cache cache = new Cache(context.getCacheDir(), cacheSize);
        OkHttpClient client = new OkHttpClient().newBuilder().cache(cache).addInterceptor(new Interceptor() {
            @Override
            public Response intercept(Chain chain) throws IOException {
                Request request = chain.request();
                if (!isNetworkAvailable(context)) {
                    int maxStale = 60 * 60 * 24 * 28; // Tolerate 4 weeks stale
                    request = request.newBuilder().header("Cache-Control", "public, only-if-cached, max-stale=" + maxStale).build();
                }
                return chain.proceed(request);
            }
        }).build();

        retrofit = new Retrofit.Builder()
                .baseUrl(ApiConfig.BASE_URL)
                .client(client)
                .addConverterFactory(GsonConverterFactory.create()).build();

        return retrofit;

    }

    private static boolean isNetworkAvailable(Context context) {
        ConnectivityManager connectivityManager = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo activeNetworkInfo = connectivityManager.getActiveNetworkInfo();
        return activeNetworkInfo != null && activeNetworkInfo.isConnected();
    }
}

File: app/src/main/java/com/timsimonhughes/atlas/network/RetrofitClientInstance.kt
package com.timsimonhughes.atlas.network

import android.content.Context
import android.net.ConnectivityManager
import okhttp3.Cache
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import okhttp3.Response
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.moshi.MoshiConverterFactory


class RetrofitClientInstance {

    private var retrofitInstance: Retrofit? = null

    fun getRetrofitInstance(context: Context?, cacheSize: Int) : Retrofit {

        retrofitInstance = Retrofit.Builder()
                .baseUrl(ApiConfig.BASE_URL)
                .addConverterFactory(MoshiConverterFactory.create())
                .client(addLoggingInterceptor())
                .client(addCache(context, cacheSize))
                .build()

        return retrofitInstance as Retrofit
    }

    private fun addCache(context: Context?, cacheSize: Int): OkHttpClient {
        val cache = context?.cacheDir?.let { Cache(it, cacheSize.toLong()) }

        val client = OkHttpClient().newBuilder().cache(cache).addInterceptor(object: Interceptor {

            override fun intercept(chain: Interceptor.Chain): Response {
                var request = chain.request()
                if (!isNetworkAvailable(context)) {
                    request = request.newBuilder().header("Cache-Control", "public, only-if-cached, max-stale=$maxStale").build()
                }
                return chain.proceed(request)
            }
        }).build()

        return client
    }

    private fun addLoggingInterceptor(): OkHttpClient {
        val loggingInterceptor = HttpLoggingInterceptor()
        loggingInterceptor.level = HttpLoggingInterceptor.Level.BODY

        val httpClientBuilder = OkHttpClient.Builder()
        httpClientBuilder.addInterceptor(loggingInterceptor)

        return httpClientBuilder.build()
    }

    private fun isNetworkAvailable(context: Context?): Boolean {
        val connectivityManager = context?.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val activeNetworkInfo = connectivityManager.activeNetworkInfo
        return activeNetworkInfo != null && activeNetworkInfo.isConnected
    }

    companion object {
        const val maxStale = 60 * 60 * 24 * 28 // Tolerate 4 weeks stale
    }

}